src/Telemetry/AzureDevOpsClient/AzureDevOpsClient.cs

```diff
@@ -155,6 +162,40 @@ public async Task<WorkItem> CreateRcaWorkItem(string project, string title, Canc
             return JsonConvert.DeserializeObject<WorkItem>(json);
         }
 
+        public async Task<string> TryGetLogContents(string logUri, CancellationToken cancellationToken)
+        {
+            for (int i = 0; i < _retryNumber; i++)
+            {
+                using (var request = new HttpRequestMessage(HttpMethod.Get, logUri))
+                {
+                    request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/plain"));
+
+                    var response = await _httpClient.SendAsync(request, cancellationToken);
+
+                    if (response.StatusCode == HttpStatusCode.TooManyRequests)
+                    {
+                        if (response.Headers.TryGetValues("Retry-After", out var values) &&
+                            int.TryParse(values.First(), out var seconds))
+                        {
+                            int randomDelay;
+                            lock (_randomNumberGenerator)
+                            {
+                                randomDelay = _randomNumberGenerator.Next(_retryDelayMin, _retryDelayMax);
+                            }
+                            await Task.Delay((int)(TimeSpan.FromSeconds(seconds).TotalMilliseconds + randomDelay), cancellationToken);
+                            continue;
+                        }
+                    }
+
+                    response.EnsureSuccessStatusCode();
```

Comment:
So we only retry for `TooManyRequests` but not other potential transient issues?

Reply:
Yes, @lpatalas and I were talking about this this morning and we think that's an ok behavior. The implementation we have in helix-services is the same too https://dev.azure.com/dnceng/internal/_search?action=contents&text=TooManyRequests&type=code&lp=code-Project&filters=ProjectFilters%7Binternal%7DRepositoryFilters%7Bdotnet-helix-machines*dotnet-helix-service%7D&pageSize=25&result=DefaultCollection/internal/dotnet-helix-service/GBmain//src/Validation/Helix.Test.Utilities/Helpers/HttpHelper.cs

Reply:
That code is used in exactly 2 places, in test. It's not a good pattern to copy.

Reply:
We should definitely try at least one for any failure.

Reply:
I added retries for all exceptions

Reply:
We use the RetryAfterHandler and exponential retry to handle these now